### DISCUSSION TITLE
Volume rendering

### DISCUSSION THREAD

**Discussion by OliviaBorgstrom** [mitsuba-renderer/mitsuba3#1101] [Q&A]
Hi, I asked this question a few days ago https://github.com/mitsuba-renderer/mitsuba3/discussions/1092, I have mostly figured out the participating media contributions. However, one issue is that in the ptracer integrator rays are currently always spawned away from the direction of the collision, which means it can't deal with volumes. 
`ray = si.spawn_ray(si.to_world(bs.wo));
si = scene->ray_intersect(ray, active);`

This is the code which spawns the rays inside of the ptracer integrator, I am assuming that bs.wo is always restricted to the top hemisphere- which is why it is never fired into the volumes. I'm wondering if anyone has any pointers on how to spawn rays which will go inside of the volumes as well. 
Thank you!

**Comment by ziyi-zhang** [comment]
Hi,
I haven't read the previous threads. But your observation is right: the current ptracer is written for surfaces without volumes. I think it was first written for the Mitsuba caustics optimization tutorial -- there is no smoke.
To add support for media, the sampling step needs to consider both volume RTE sampling (usually in a delta tracking manner) and the surface scattering. The volume RTE sampling would give a "medium interaction" (`mei`), from which we sample the phase function. This is analogous to sampling a BSDF from a "surface interaction" (`si`), but the two logics are separate. An example can be found in `volpath`.
Best

**Reply by OliviaBorgstrom** [reply]
Hi @ziyi-zhang, Thank you for your response. I think I understand what you mean, I have looked into the volpath integrator and what I currently have is an isolated function which calculates the medium contributions for a ray. What I am not sure of is how to get rays to fire into the volume so that they can pick up the medium contributions. From what I currently understand about ptracer, it traces a single ray and at every bounce it spawns a different ray one in some(? - Is it random ?) direction away from the collision.

My current plan is to use the medium contribution function which I have and add the result of this to the throughput in the trace_light_ray function - but to have any meaningful contribution happen I need to figure out how I can allow rays to be shot into the volumes. This is what I think(?) my problem is, but if I'm missing anything here please let me know. Thank you!

TLDR:
- Does ptracer spawn a ray in a random direction at every bounce
- How do I change the ray spawning to send rays into the volume (if im correct that this is my issue)

**Reply by ziyi-zhang** [reply]
Hi,

> Does ptracer spawn a ray in a random direction at every bounce

(Not unique to ptracer) No. We sample proportional to a certain distribution, and the sampling density must be divided out. Uniform distribution is one choice, but we prefer importance sampling the BSDF/phase function to reduce MC variance.

> How do I change the ray spawning to send rays into the volume?

If you are referring to camera ray spawning,  we never do that in ptracer. `ptracer`, or light tracer, simulates the process of light photons/particles generating from an emitter, bouncing in the scene, and eventually hitting the camera film. Shooting rays from the camera to the scene would lead to bidirectional integrators.  
If you are referring to "next event estimation", then it's the same as what we did for surfaces -- call "sensor.sample_direction()". But be careful the transmittance from the current mei to the sensor needs to be considered.

Aside: if you do not need AD, you can take a look at mitsuba 0.6, where the ptracer supports volume. 
best

**Reply by OliviaBorgstrom** [reply]
Hi, 

> (Not unique to ptracer) No. We sample proportional to a certain distribution, and the sampling density must be divided out. Uniform distribution is one choice, but we prefer importance sampling the BSDF/phase function to reduce MC variance.

This makes sense thank you 

> If you are referring to "next event estimation", then it's the same as what we did for surfaces -- call "sensor.sample_direction()". But be careful the transmittance from the current mei to the sensor needs to be considered.

Do you think you could explain what next event estimation means? 

<img width="568" alt="image" src="https://github.com/mitsuba-renderer/mitsuba3/assets/90727726/6df7d5c7-9f9f-43de-8117-7b0193b34ed0">

I was mostly referring to these lines in the ptracer code, what is the purpose of them?

Also thank you for the pointer to mitsuba 0.6 - I do not need AD, however I do still need to use mitsuba 3 due to previous work on this project. However, I may be able to learn something from the implementation.

**Reply by merlinND** [reply]
Hello @OliviaBorgstrom,

> Do you think you could explain what next event estimation means?

If you haven´t already, I would recommend reading through a few chapters of PBRT which should explain things clearly and make the Mitsuba code easier to follow: https://pbr-book.org/4ed/Light_Transport_I_Surface_Reflection/Path_Tracing
